api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -131,34 +131,33 @@ def __next__(self) -> ObsDatum:
         while len(obs) < self.batch_size:
             try:
                 obs_partial, X_partial = self._read_partial_torch_batch(self.batch_size)
-                if X is None:
-                    obs = obs_partial
-                    X = X_partial
-                else:
-                    obs = pd.concat([obs, obs_partial], axis=0)
-                    X = sparse.vstack([X, X_partial])
-
+                obs = pd.concat([obs, obs_partial], axis=0)
+                X = sparse.vstack([X, X_partial])
             except StopIteration:
                 break
 
         if len(obs) == 0:
             raise StopIteration
 
+        assert obs["soma_joinid"].is_unique
+
         obs_encoded = pd.DataFrame(
             data={"soma_joinid": obs.soma_joinid}, columns=obs.columns, dtype=np.int32, index=obs.index
         )
         for col, enc in self.encoders.items():
             obs_encoded[col] = enc.transform(obs[col])
 
-        obs_tensor = torch.Tensor(obs_encoded.to_numpy()).int()
+        # `as_tensor()` avoids copying the numpy array data
+        obs_tensor = torch.from_numpy(obs_encoded.to_numpy(dtype=np.int32))
```

Comment:
This is the bug fix

Reply:
I am concerned this may have another bug - it casts the soma_joinid int64 to int32 on line 147, which will lose part of the joinid value.  If these are _really_ soma_joinid values, they they can have any value in the int64 domain including the high 4 bytes.  The census builder currently assigns them in the range [0, n_cells), so it is not yet likely to fail as our collections are < 2^31-1 in size, but it may/will at some point.  And other datasets are free to assign joinids as any legal int64.

If the values are not soma_joinids, but are reindexed to [0, n), then the code is confusing as it calls them "joinids" when they are something different.  And even in this case, we will have collections with > 2^31-1 cells at some point - might as well handle them.

So, full suggestion:  either:
* obs_encoded should not drop top 4 bytes of the joinid, and should create int64 dataframe/tensor
* or, needs to error out when there are more than 2^31-1 cells, AND reindex the joinids down to int32

Unrelated side note: as I read the pytorch docs, they imply that constructors are never the preferred way to create a Tensor, and it should be done with `torch.tensor()`, `torch.from_numpy()`, etc.  Never explicitly stated though, so this may be incorrect.

Reply:
I think this is a valid point but since the census is strongly biased towards cell data, it is also unrealistic that we'll ever have more than `2^31-1` cells within the current iteration. This is roughly 50 times what we have now and might exceed the total number of cells ever sequenced in single cell experiments (although this is purely a guess). That said, this is the same reasoning that led to the Y2K problem (or even the R limitation for sparse matrices), so if using int64 is feasible/doesn't add a lot of work, it might be worth it...

Reply:
I can think of several reasons we might have joinids > 2^31-1 (e.g., we start a policy of not reusing them, we have a sparse range for some other reason, etc).

Bottom line: the code looks wrong or unsafe at best.  Minimally, I would add explicit assertions/error checks, but as joinids are not offsets, it is a bit of a cringe to say they are guaranteed to remain within a range defined by the number of cells.

Reply:
I'll bring the precision in line with the SOMA spec (int64).  In recent TileDB-SOMA discussions, I thought we were moving to int32 for max shape of experiments, but may have misunderstood. In any case int64 is safest for now, agreed.

Reply:
I also found that `Torch()` doesn't seem to be the Right Way to create tensors, though the library didn't complainat me unfortunately. As class constructor doesn't even take `dtype` param, I was properly coerced into using the proper factory methods.

Reply:
No, we are definitely _not_ moving to int32 joinids - was never on the table.  I think you may be conflating several issues (e.g., capacity vs utilized ids).

There is a proposal to:
* set capacity by default to a small number in `soma.io`.  This does not affect the Census as it doesn't use that code
* introduce a means to ask "how many coordiates are used", but this also has no effect on the domain from which legal values can be chosen.

These are completely separate issues in reality - joindids are not offset indices

Reply:
fixed